### DISCUSSION TITLE
Deploying the backend to Azure

### DISCUSSION THREAD

**Discussion by decapent** [pnp/generator-teams#210] [Q&A] [answered]
I have completed my first search-based messaging extension. Now all works fine in a local scenario by leveraging ngrok. I am at the point that I need to deploy to the client's environment and would like to know what is the next step.

My first approach would be an AppService in Azure. What are my options ?

Also, what are the best practices regarding environment specific configurations contained in the .env file ?

**Comment by wictorwilen** [comment, marked as answer]
Good question. There's multiple ways of doing this - and if Azure is your target cloud then running it in an Azure App service/web application is most likely your best option - and it's fairly easy to set up. I previously wrote a post on how I do this, for my projects, with Azure DevOps: https://www.wictorwilen.se/blog/deploying-yo-teams-and-node-apps/. The pipeline could easily be replicated for other CI/CD services. 
When it comes to the environment variables, having a `.env` file is great for dev but I don't recommend it in production. Use either the Azure App Service configuration or even better Azure App Configuration (for the latter I created a small utility package that allows you to easily just use `process.env.*` with App Config. See https://www.npmjs.com/package/azure-env-app-configuration)

**Comment by wictorwilen** [comment]
Another option that you can use to deploy the solution is to user Containers and Docker. The scaffolded project already contains a `Dockerfile` for this. Then you can deploy the Docker image to an Azure App Service, Docker or Kubernetes.

**Comment by decapent** [comment]
Thank you very much, this is exactly what I needed 👍🏻

**Comment by decapent** [comment]
I have packaged my application as per the provided blog post. I have deployed to an Azure App Service and recreated the environment variables. However, the application keeps getting an HTTP error 500 from the bot framework when trying to POST to /api/messages.

Local Setup works, I have tested the GraphConnection made by the bot, updated the bot messaging endpoint to target my App Service host name .

The following is the error I keep getting:

BotFrameworkAdapter.processActivity(): 500 ERROR
Error at BotFrameworkAdapter.<anonymous>

Any idea what might be happening ?

**Reply by wictorwilen** [reply]
Really hard to tell what it is, any chance you can log the full HTTP 500 response, including the error message/stack?

**Reply by decapent** [reply]
Of course, it is my bad for not providing any additional information

The error arrives on the initialRun of the bot.

The Console
![image](https://user-images.githubusercontent.com/6863827/115738410-07043600-a35b-11eb-9907-96b35dc12d01.png)

`2021-04-22T15:04:52.199Z InputExtensionViewModel: Failed to complete command request for loadDefaultItems. {"statusCode":502,"headers":{"cache-control":"no-store, must-revalidate, no-cache","content-length":"87","content-type":"application/json; charset=utf-8","contextid":"tcid=0,server=msgapi-production-wus-azsc4-2-222,cv=ECffe0F4g0a/ewd93bTWJA.0","pragma":"no-cache"},"errorMessage":"<BotError>Bot returned unsuccessful status code NotFound","errorCode":1008,"request":{}}`

The Network
![image](https://user-images.githubusercontent.com/6863827/115738717-4e8ac200-a35b-11eb-9d52-af9a608d80ff.png)

Also I am using the same app registrations (will be changed for prod) the only difference is that that the code has been deployed to an Azure App Service. The App service is up and running and sends back a page when poking the endpoint.

![image](https://user-images.githubusercontent.com/6863827/115739163-ad503b80-a35b-11eb-8f36-c9383bc46ed1.png)

And I have updated the bot messaging endpoint accordingly :

![image](https://user-images.githubusercontent.com/6863827/115739629-15068680-a35c-11eb-81c8-36be43f7838c.png)

**Reply by wictorwilen** [reply]
I would check 
1) in the bot resource in Azure, use the Web Chat to test it
2) Check the app service logs (ensure that you add `DEBUG=msteams` to app settings and see if there's an incoming request

**Reply by decapent** [reply]
Thank you, it did provide some insights. Turns out that an underlaying exception was thrown and pointed out that the bot wasn't registered properly which lead me to this article that I will leave here for posterity.

[http://thewindowsupdate.com/2019/08/19/how-to-create-an-app-registration-in-azure-ad-for-a-bot-solution/](http://thewindowsupdate.com/2019/08/19/how-to-create-an-app-registration-in-azure-ad-for-a-bot-solution/)

Thanks a lot for your continuous support with this issue Wictor :)
